kstat.rs

```diff
@@ -0,0 +1,313 @@
+    l: c_long,
+    ul: c_ulong,
+    ui32: u32,
+}
+
+#[repr(C)]
+struct KstatNamed {
+    name: [c_char; KSTAT_STRLEN],
+    data_type: c_uchar,
+    value: KstatValue,
+}
+
+extern "C" {
+    fn kstat_open() -> *mut KstatCtl;
+    fn kstat_close(kc: *mut KstatCtl) -> c_int;
+    fn kstat_lookup(kc: *mut KstatCtl, module: *const c_char,
+        instance: c_int, name: *const c_char) -> *mut Kstat;
+    fn kstat_read(kc: *mut KstatCtl, ksp: *mut Kstat, buf: *mut c_void)
+        -> c_int;
+    fn kstat_data_lookup(ksp: *mut Kstat, name: *const c_char) -> *mut c_void;
+}
+
+/// Minimal wrapper around libkstat(3LIB) on illumos and Solaris systems.
+struct KstatWrapper {
+    kc: *mut KstatCtl,
+    ksp: *mut Kstat,
+    stepping: bool,
+}
+
+/// Turn an optional string reference into an optional owned C string.
+fn cstr(s: Option<&str>) -> Option<std::ffi::CString> {
+    s.map_or(None, |s| Some(CString::new(s.to_owned()).unwrap()))
+}
+
+/// Turn an optional C string into a (const char *) for Some, or NULL for None.
+fn cstrp(p: &Option<CString>) -> *const c_char {
+    if let Some(p) = p.as_ref() {
+        p.as_ptr()
+    } else {
+        null()
+    }
+}
+
+impl KstatWrapper {
+    fn open() -> Result<Self> {
+        let kc = unsafe { kstat_open() };
+        if kc == null_mut() {
+            return Err("kstat_open(3KSTAT) failed".into());
+        }
+        Ok(KstatWrapper {
+            kc: kc,
+            ksp: null_mut(),
+            stepping: false,
+        })
+    }
+
+    /// Call kstat_lookup(3KSTAT) and store the result, if there is a match.
+    fn lookup(&mut self, module: Option<&str>, name: Option<&str>) {
+        let module = cstr(module);
+        let name = cstr(name);
+
+        unsafe {
+            self.ksp = kstat_lookup(self.kc, cstrp(&module), -1, cstrp(&name));
+        }
+
+        self.stepping = false;
+    }
+
+    /// Call once to start iterating, and then repeatedly for each additional
+    /// kstat in the chain.  Returns false once there are no more kstat entries.
+    fn step(&mut self) -> bool {
+        if !self.stepping {
+            self.stepping = true;
+        } else {
+            self.ksp = unsafe { (*self.ksp).ks_next };
+        }
+
+        if self.ksp == null_mut() {
+            self.stepping = false;
+            false
+        } else {
+            true
+        }
+    }
+
+    /// Return the module name of the current kstat.
+    fn module(&self) -> Option<String> {
+        if self.ksp == null_mut() {
+            None
+        } else {
+            Some(unsafe { (*self.ksp).module() })
+        }
+    }
+
+    /// Return the name of the current kstat.
+    fn name(&self) -> Option<String> {
+        if self.ksp == null_mut() {
+            None
+        } else {
+            Some(unsafe { (*self.ksp).name() })
+        }
+    }
+
+    /// Look up a named kstat value.  For internal use by typed accessors.
+    fn data_value(&self, statistic: &str) -> Option<*const KstatNamed> {
```

Comment:
So, if the kstat stuff gets into its own module (or all clients remain correct forever), I think you could return `Option<&KstatNamed>`.

This would also prevent a caller from trying to drop `self` while holding this pointer, which (I gather) would be Bad.

Reply:
I couldn't figure out a good way to return an `Option<&KstatNamed>`, and have instead opted for `Option<NonNull<KstatNamed>>`.